spanner/stats.go

```diff
@@ -167,3 +210,108 @@ func EnableStatViews() error {
+	GFELatencyMetricsEnabled = true
+	return view.Register(GFELatencyView)
+}
+
+// EnableGfeHeaderMissingCountView enables GFEHeaderMissingCount metric
+func EnableGfeHeaderMissingCountView() error {
+	GFELatencyMetricsEnabled = true
+	return view.Register(GFEHeaderMissingCountView)
+}
+
+// EnableGfeLatencyAndHeaderMissingCountViews enables GFEHeaderMissingCount and GFELatency metric
+func EnableGfeLatencyAndHeaderMissingCountViews() error {
+	GFELatencyMetricsEnabled = true
+	return view.Register(
+		GFELatencyView,
+		GFEHeaderMissingCountView,
+	)
+}
+
+// DisableGfeLatencyAndHeaderMissingCountViews disables GFEHeaderMissingCount and GFELatency metric
+func DisableGfeLatencyAndHeaderMissingCountViews() {
+	GFELatencyMetricsEnabled = false
+	view.Unregister(
+		GFELatencyView,
+		GFEHeaderMissingCountView,
+	)
+}
+
+func captureGFELatencyStats(ctx context.Context, md metadata.MD, keyMethod string) error {
+	if len(md.Get("server-timing")) == 0 {
+		recordStat(ctx, GFEHeaderMissingCount, 1)
+		return nil
+	}
+	serverTiming := md.Get("server-timing")[0]
+	gfeLatency, err := strconv.Atoi(strings.TrimPrefix(serverTiming, "gfet4t7; dur="))
+	if !strings.HasPrefix(serverTiming, "gfet4t7; dur=") || err != nil {
+		return err
+	}
+	// Record GFE latency with OpenCensus.
+	ctx = tag.NewContext(ctx, tag.FromContext(ctx))
+	ctx, err = tag.New(ctx, tag.Insert(tagKeyMethod, keyMethod))
+	if err != nil {
+		return err
+	}
+	recordStat(ctx, GFELatency, int64(gfeLatency))
+	return nil
+}
+
+func checkCommonTagsGFELatency(t *testing.T, m map[tag.Key]string) {
+	// We only check prefix because client ID increases if we create
+	// multiple clients for the same database.
+	if !strings.HasPrefix(m[tagKeyClientID], "client") {
+		t.Fatalf("Incorrect client ID: %v", m[tagKeyClientID])
+	}
+	if !strings.HasPrefix(m[tagKeyInstance], "gotest") {
+		t.Fatalf("Incorrect instance ID: %v", m[tagKeyInstance])
+	}
+	if !strings.HasPrefix(m[tagKeyDatabase], "gotest") {
+		t.Fatalf("Incorrect database ID: %v", m[tagKeyDatabase])
+	}
+	if m[tagKeyLibVersion] != version.Repo {
+		t.Fatalf("Incorrect library version: %v", m[tagKeyLibVersion])
+	}
+}
+
+func createContextAndCaptureGFELatencyMetrics(ctx context.Context, ct *CommonTags, md metadata.MD, keyMethod string) error {
+	var ctxGFE, err = tag.New(ctx,
+		tag.Upsert(tagKeyClientID, ct.clientID),
+		tag.Upsert(tagKeyDatabase, ct.database),
+		tag.Upsert(tagKeyInstance, ct.instance),
+		tag.Upsert(tagKeyLibVersion, ct.libVersion),
+	)
+	if err != nil {
+		return err
+	}
+	return captureGFELatencyStats(ctxGFE, md, keyMethod)
+}
+
+func getCommonTags(sc *sessionClient) *CommonTags {
+	_, instance, database, err := parseDatabaseName(sc.database)
+	if err != nil {
+		return nil
+	}
+	return &CommonTags{
+		clientID:   sc.id,
+		database:   database,
+		instance:   instance,
+		libVersion: version.Repo,
+	}
+}
+
+// CommonTags are common key-value pairs of data associated with the GFELatency measure
+type CommonTags struct {
```

Comment:
nit: Could this struct be unexported (that is; `commonTags` instead of `CommonTags`)?